### PR TITLE
feat(DTFS2-6182): replace GraphQL TFM endpoints with REST

### PR DESCRIPTION
**Pull request by francescastocco**
## Introduction
Migrating the queryFacility GraphQL endpoint to a REST endpoint.
This is the start of a long-lived branch to replace graphql. looking at using the existing graphql reducers -> copy/paste to graphwl mappings as it will most likely be rewritten

## Resolution
- Updated frontend to make a REST API call
- Updated backend route & controller to REST
- Using the existing graphql reducers which have been copy/pasted to a new `graphql-mappings` folder as it's likely that these will be rewritten soon

## To Do
- Return same response body as graphql

## Misc
- Updated error messages typo in DTFS Central API

### CODE REVIEW

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityType.api-test.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+      const result = mapFacilityType(mockContingentFacility);
+
+      const expected = `${CONSTANTS.FACILITIES.FACILITY_PRODUCT_NAME.CONTINGENT} facility`;
+
+      expect(result).toEqual(expected);
+    });
+  });
+
+  it('should return null', () => {
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityType.api-test.js:68]: needs a describe block

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityStage.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+const CONSTANTS = require('../../../../constants');
+
+const mapFacilityStage = (facilityStage) => {
+  let mapped;
+
+  // NOTE: we currently have two different data sources.
+  // - Bond & Loan facilities have facilityStage string of 'issued' or 'unissued'.
+  // - Cash & Contingent facilities have hasBeenIssued boolean.
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityStage.js:8]: I feel we could reduce some of these comments by assigning our or checks to constants?

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityStage.api-test.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  describe('when facilityStage is `Unconditional`', () => {
+    it('should return facilityStage as `Issued`', () => {
+      const result = mapFacilityStage('Unconditional');
+      expect(result).toEqual('Issued');
+    });
+  });
+
+  describe('when facilityStage is true boolean', () => {
+    it('should return facilityStage as `Commitment`', () => {
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityStage.api-test.js:19]: Test doesn't match with name

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacility.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+  const {
+    type,
+    value,
+    facilityStage,
+    guaranteeFeePayableByBank,
+    currency,
+  } = facility;
+
+  facility.ukefFacilityType = type;
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacility.js:33]: Again -- not sure I like reassigning. here :) 

Also -- Seems a tad odd we're reassigning `ukefFacilityType` to be `type` (and without mapping?)

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacility.api-test.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  const mockFacilityValue = '12345.00';
+  const mockType = 'Bond';
+  const mockFacilityStage = 'Unissued';
+
+  const mockFacility = {
+    _id: '12345678',
+    dealId: '100200300',
+    ukefFacilityId: '0040004833',
+    type: mockType,
+    ukefFacilityType: mockType,
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacility.api-test.js:48]: how sure are we that these mock types are the same (type/ukefFacilityType)?

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilities.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+const mapFacility = require('./mapFacility');
+const mapFacilityTfm = require('./mapFacilityTfm');
+
+const mapFacilities = (facilities, dealDetails, dealTfm) => {
+  // Map facilities if only they exists
+  if (facilities?.length) {
+    const mappedFacility = facilities.map((f) => ({
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilities.js:7]: `facility`

**Review by AlexBramhill — COMMENTED** [review]
General comment -- We have `formattedNumber` for every time we wish to add 2dp to a number -- We should try and use that I think :) 

-- will rereview once we have an idea of how many of these mappings are needed. Only half way through

**Review by AlexBramhill — COMMENTED** [review]
reviewed non-mapping files. Mapping files still to be reviewed.

**Comment by AlexBramhill** [comment]
General comments:
A lot of `return {}` where it might be better returning something else in `tfm-ui`

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -8,7 +8,7 @@ run-name: Executing test QA on ${{ github.repository }} 🚀
 
 on:
   pull_request:
-    branches: [main]
+    branches: [main, feat/DTFS2-6182/replace-graphql-query-facility-in-tfm-with-rest]
```
Comment (AlexBramhill) [inline .github/workflows/test.yml:11]: Remember to revert 

```suggestion
    branches: [main]
```

File: .github/workflows/lint.yml [hunk 8]
```diff
@@ -8,7 +8,7 @@ run-name: Executing lint QA on ${{ github.repository }} 🚀
 
 on:
   pull_request:
-    branches: [main]
+    branches: [main, feat/DTFS2-6182/replace-graphql-query-facility-in-tfm-with-rest]
```
Comment (AlexBramhill) [inline .github/workflows/lint.yml:11]: Remember to revert
```suggestion
    branches: [main]
```

File: trade-finance-manager-api/src/v1/graphql-mappings/deal-with-amendment-bss.api-test.js [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    expect(amendmentResult.dealSnapshot.facilities[0].facilitySnapshot.value).toEqual(amendedValue);
+    expect(amendmentResult.dealSnapshot.facilities[0].facilitySnapshot.dates.coverEndDate).toEqual(amendedCoverEndDate);
+    expect(amendmentResult.dealSnapshot.facilities[0].facilitySnapshot.dates.tenor).toEqual(amendedTenor);
+  });
+
+  it('should return updated deal with completed amendment with 1 changes if 1 accepted and 1 rejected', () => {
+    MOCK_DEAL_AIN_SUBMITTED.loanTransactions.items[0]['requestedCoverStartDate-year'] = '2022';
+    MOCK_DEAL_AIN_SUBMITTED.loanTransactions.items[0]['requestedCoverStartDate-month'] = '01';
+    MOCK_DEAL_AIN_SUBMITTED.loanTransactions.items[0]['requestedCoverStartDate-day'] = '01';
+    mockAmendment.ukefDecision.coverEndDate = AMENDMENT_UW_DECISION.DECLINED;
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/graphql-mappings/deal-with-amendment-bss.api-test.js:115]: I'm leaving these changes as not done for now -- This is a relatively large debt that should be tackled in due course

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/gef-deal/mapGefSubmissionDetails.api-test.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  const mockDeal = {
+    _id: MOCK_GEF_DEAL._id,
+    dealSnapshot: {
+      ...MOCK_GEF_DEAL,
+      facilities: [],
+    },
+    tfm: {},
+  };
+
+  it('should return mapped submission details', () => {
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/graphql-mappings/mappings/gef-deal/mapGefSubmissionDetails.api-test.js:14]: I'm leaving these changes as not done for now -- This is a relatively large debt that should be tackled in due course

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacility.api-test.js (hunk 5 above)
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacility.api-test.js:48]: As these are existing and due to the number of changes in this PR, we shall leave this as is

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityStage.js (hunk 2 above)
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityStage.js:8]: As these are existing and due to the number of changes in this PR, we shall leave this as is

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacility.js (hunk 4 above)
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacility.js:33]: As these are existing and due to the number of changes in this PR, we shall leave this as is

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityValue.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+      if (latestAmendmentTFM?.value) {
+        const valueInGBP = calculateNewFacilityValue(exchangeRate, latestAmendmentTFM.value);
+        return `${CURRENCY.GBP} ${formattedNumber(valueInGBP)}`;
+      }
+    }
+
+    if (facilityTfm?.facilityValueInGBP) {
+      const { facilityValueInGBP } = facilityTfm;
+
+      if (currencyId !== CURRENCY.GBP) {
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapFacilityValue.js:25]: As these are existing and due to the number of changes in this PR, we shall leave this as is

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapPremiumTotals.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+const mapPremiumTotals = (premiumSchedule) => {
+  if (premiumSchedule) {
+    return premiumSchedule.map((s) => s.income).reduce((a, b) => a + b, 0).toFixed(2);
+  }
+
+  return '0';
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapPremiumTotals.js:6]: As these are existing and due to the number of changes in this PR, we shall leave this as is

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapTenorDate.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  // Un-issued facility
+  if (!hasBeenIssued) {
+    return `${months} ${monthString(months)}`;
+  }
+
+  // Issued facility
+  if (hasBeenIssued) {
+    return `${months} ${monthString(months)}`;
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/graphql-mappings/mappings/facilities/mapTenorDate.js:38]: As these are existing and due to the number of changes in this PR, we shall leave this as is

File: trade-finance-manager-api/src/v1/graphql-mappings/mappings/gef-facilities/mapGefFacility.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+      facilityProduct: facilitySnapshot.facilityProduct,
+      facilityStage: facilitySnapshot.facilityStage,
+      hasBeenIssued: facilitySnapshot.hasBeenIssued,
+      type: mapFacilityType(facilitySnapshot),
+      currency: currency.id,
+      facilityValueExportCurrency: mapFacilityValueExportCurrency(facility),
+      value: mapFacilityValue(currency.id, formattedFacilityValue, facility),
+      feeType,
+      feeFrequency,
+      guaranteeFeePayableToUkef: mapGuaranteeFeePayableToUkef(guaranteeFee),
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/graphql-mappings/mappings/gef-facilities/mapGefFacility.js:60]: As these are existing and due to the number of changes in this PR, we shall leave this as is

**Review by lmarrai-sw — APPROVED** [review]
All the changes in this PR have been reviewed individually (save for the 2 latest commits which I have reviewed and approve of)

**Pull request merged**